packages/app-headless-cms/src/admin/plugins/fieldRenderers/dynamicZone/TemplateGallery.tsx

```diff
@@ -67,29 +67,23 @@ const Gallery = makeDecoratable("TemplateGalley", (props: TemplateGalleryProps)
     );
 });
 
-function withDecoratorFactory<T extends DecoratableComponent>(Component: T) {
-    return Object.assign(Component, {
-        createDecorator: createDecoratorFactory<{ modelIds?: string[] }>()(
-            Component,
-            decoratorProps => {
-                const { model } = useModel();
+export type ShouldRender = { modelIds?: string[] };
 
-                if (
-                    decoratorProps?.modelIds?.length &&
-                    !decoratorProps.modelIds.includes(model.modelId)
-                ) {
-                    return false;
-                }
+function withShouldRender<T extends DecoratableComponent>(Component: T) {
+    return withDecoratorFactory<ShouldRender>()(Component, decoratorProps => {
+        const { model } = useModel();
 
-                return true;
-            }
-        )
+        if (decoratorProps?.modelIds?.length && !decoratorProps.modelIds.includes(model.modelId)) {
+            return false;
+        }
+
+        return true;
     });
 }
 
-export const TemplateGallery = Object.assign(withDecoratorFactory(Gallery), {
-    Container: withDecoratorFactory(GalleryContainer),
-    List: withDecoratorFactory(GalleryList),
-    Item: withDecoratorFactory(TemplateItem),
-    Close: withDecoratorFactory(CloseGallery)
+export const TemplateGallery = Object.assign(withShouldRender(Gallery), {
```

Comment:
Can you explain this to me? Why are we wrapping all components with `withShouldRender` and not only `Gallery`? 

I probably don't understand what's going on here.

Reply:
We're wrapping each component, because they're all decoratable, and this `withShouldRender` attaches a conditional decorator, which optionally takes a `modelIds` props, so you can control on which models that component will be decorated.

TemplateGallery consists of many components, Container, List, Item, Close, and they're all decoratable.

Reply:
Ah I understand now, tnx!